File: server/src/main/java/com/example/demo/controllers/RecipeController.java
package com.example.demo.controllers;

import com.example.demo.dto.NewRecipeCommentDTO;
import com.example.demo.dto.RecipeRatingDTO;
import com.example.demo.models.Recipe;
import com.example.demo.models.RecipeComment;
import com.example.demo.models.RecipeRating;
import com.example.demo.services.RecipeService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping("/api/recipes")
@CrossOrigin
public class RecipeController {

    @Autowired
    RecipeService recipeService;

    @GetMapping
    public ResponseEntity<Page<Recipe>> getAll(
            @RequestParam(defaultValue = "", required = false) String title,
            @RequestParam(defaultValue = "1", required = false) int page,
            @RequestParam(defaultValue = "12", required = false) int pageSize,
            @RequestParam(required = false) String sortby) {
        Page<Recipe> recipes = recipeService.getBySearchParams(title, page, pageSize, sortby);
        return ResponseEntity.ok().body(recipes);
    }

    @PostMapping
    public ResponseEntity<Recipe> create(
            @RequestHeader("Authorization") String bearerToken,
            @RequestPart(value = "document") String recipeJson,
            @RequestPart(value = "image", required = false) MultipartFile backgroundImage) {
        return ResponseEntity
            .status(HttpStatus.CREATED)
            .body(recipeService.create(recipeJson, backgroundImage, bearerToken));
    }

    @GetMapping("/{id}")
    public ResponseEntity<Recipe> findById(@PathVariable Long id) {
        return ResponseEntity.ok(recipeService.findById(id));
    }

    @PostMapping("/{recipeId}/comments")
    public ResponseEntity<RecipeComment> addComment(
            @RequestHeader("Authorization") String bearerToken,
            @PathVariable Long recipeId,
            @RequestBody NewRecipeCommentDTO comment) {
        return ResponseEntity
            .status(HttpStatus.CREATED)
            .body(recipeService.addComment(recipeId, comment, bearerToken));
    }

    @PostMapping("/{recipeId}/ratings")
    public ResponseEntity<RecipeRating> addRating(
            @RequestHeader("Authorization") String bearerToken,
            @PathVariable Long recipeId,
            @RequestBody RecipeRatingDTO ratingDTO) {
        return ResponseEntity
            .status(HttpStatus.CREATED)
            .body(recipeService.addRating(recipeId, ratingDTO, bearerToken));
    }
}

File: server/src/main/kotlin/com/example/demo/recipe/RecipeController.kt
package com.example.demo.recipe

import com.example.demo.ApiPath
import com.example.demo.domain.PageResult
import com.example.demo.recipe.domain.CreateRecipeCommentRequest
import com.example.demo.recipe.domain.CreateRecipeRatingRequest
import com.example.demo.recipe.domain.CreateRecipeRequest
import com.example.demo.recipe.domain.GetRecipeResponse
import com.example.demo.recipe.domain.RecipeListItemResponse
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile

@RestController
@RequestMapping(ApiPath.RECIPES_API)
class RecipeController(private val recipeService: RecipeService) {

    @GetMapping
    fun getAll(
        @RequestParam(name = "title", defaultValue = "", required = false) recipeTitle: String,
        @RequestParam(name = "category", required = false) category: String?,
        @RequestParam(name = "page", defaultValue = "1", required = false) page: Int,
        @RequestParam(name = "page_size", defaultValue = "12", required = false) pageSize: Int,
        @RequestParam(name = "sort_by", defaultValue = "date_desc", required = false) sortBy: String,
    ): ResponseEntity<PageResult<RecipeListItemResponse>> {
        val normalizedCategory = category?.trim()?.lowercase()?.takeIf { it.isNotEmpty() && it != "all" }

        return ResponseEntity.ok(
            recipeService.getAll(
                recipeTitle = recipeTitle,
                category = normalizedCategory,
                page = page,
                pageSize = pageSize,
                sortBy = sortBy,
            )
        )
    }

    @GetMapping("/{id}")
    fun getRecipeById(@PathVariable id: Int): ResponseEntity<GetRecipeResponse> =
        ResponseEntity.ok(recipeService.findById(id))

    @PostMapping
    fun createRecipe(@RequestBody createRecipeDTO: CreateRecipeRequest): ResponseEntity<GetRecipeResponse> =
        ResponseEntity.status(HttpStatus.CREATED).body(recipeService.createRecipe(createRecipeDTO))

    @PostMapping("/{id}/comments")
    fun addComment(
        @PathVariable id: Int,
        @RequestBody commentDto: CreateRecipeCommentRequest,
    ): ResponseEntity<GetRecipeResponse> =
        ResponseEntity.status(HttpStatus.CREATED).body(recipeService.addComment(id, commentDto))

    @PostMapping("/{id}/ratings")
    fun addRating(
        @PathVariable id: Int,
        @RequestBody rating: CreateRecipeRatingRequest,
    ): ResponseEntity<GetRecipeResponse> =
        ResponseEntity.status(HttpStatus.CREATED).body(recipeService.createRecipeRating(id, rating))

    @PostMapping("/{id}/image", consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    fun uploadImage(
        @PathVariable id: Int,
        @RequestPart("image") image: MultipartFile,
    ): ResponseEntity<GetRecipeResponse> = ResponseEntity.ok(recipeService.uploadRecipeImage(id, image))

    @PutMapping("/{id}/ratings")
    fun updateRating(
        @PathVariable id: Int,
        @RequestBody rating: CreateRecipeRatingRequest,
    ): ResponseEntity<GetRecipeResponse> = ResponseEntity.ok(recipeService.updateRating(id, rating))
}
